File: cmd/vex/test/mus_test.c
// Sequencer conformance for mus.h: drives the tracker with a stub tone()
// that records calls, pinning trigger/rest/off semantics, the sustain
// default (speed*2), per-note volume overrides with peak-at-vol, the
// end-of-song stop, and out-of-range instrument bounds checks.

#include <stdio.h>
#include <string.h>

// Redirect tone() to a recorder before mus.h (via vex.h) declares it.
#define tone mus_record_tone
#include "mus.h"
#undef tone

#define MAX_CALLS 64
static struct {
    int freq, dur, vol, flags;
} g_calls[MAX_CALLS];
static int g_ncalls = 0;

void mus_record_tone(int freq, int duration, int volume, int flags) {
    if (g_ncalls < MAX_CALLS) {
        g_calls[g_ncalls].freq = freq;
        g_calls[g_ncalls].dur = duration;
        g_calls[g_ncalls].vol = volume;
        g_calls[g_ncalls].flags = flags;
        g_ncalls++;
    }
}

static int failures = 0;

#define CHECK(desc, cond) do {                                              \
        if (!(cond)) {                                                      \
            fprintf(stderr, "FAIL: %s\n", desc);                            \
            failures++;                                                     \
        }                                                                   \
    } while (0)

static const MusInst INSTS[] = {
    {VEX_TONE_PULSE, VEX_TONE_MODE0, 0, 1, 0, 4, 80, 0},
    {VEX_TONE_NOISE, 0,              0, 0, 5, 2, 40, 0},
    {VEX_TONE_PULSE, VEX_TONE_MODE0, 0, 0, 7, 4, 60, 0},
    {VEX_TONE_PULSE, VEX_TONE_MODE0, 0, 0, MUS_SUSTAIN_HOLD, 3, 50, 0},
    {VEX_TONE_PULSE, VEX_TONE_MODE0, 0, 0, 0, 1, 0, 0},
};

static const MusEvent EVENTS[4 * MUS_CHANNELS] = {
    {69, 1, 0}, {67, 1, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0},
    {MUS_REST, 0, 0}, {MUS_OFF, 0, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0},
    {129, 3, 33}, {60, 2, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0},
    {62, 9, 0}, {MUS_OFF, 0, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0},
};

static const MusPat PAT = {4, 2, EVENTS};
static const MusPat *const PATS[] = {&PAT};
static const unsigned char ORDERS[] = {0};
static const MusSong SONG = {5, 1, 1, 0xFF, INSTS, PATS, ORDERS};

static const MusEvent ARP_EVENTS[2 * MUS_CHANNELS] = {
    {129, 3, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0},
    {MUS_REST, 0, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0},
};
static const MusPat ARP_PAT = {2, 7, ARP_EVENTS};
static const MusPat *const ARP_PATS[] = {&ARP_PAT};
static const MusSong ARP_SONG = {5, 1, 1, 0xFF, INSTS, ARP_PATS, ORDERS};

static const MusEvent HOLD_EVENTS[2 * MUS_CHANNELS] = {
    {60, 4, 0}, {60, 3, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0},
    {MUS_OFF, 0, 0}, {60, 5, 0}, {MUS_REST, 0, 0}, {MUS_REST, 0, 0},
};
static const MusPat HOLD_PAT = {2, 1, HOLD_EVENTS};
static const MusPat *const HOLD_PATS[] = {&HOLD_PAT};
static const MusSong HOLD_SONG = {5, 1, 1, 0xFF, INSTS, HOLD_PATS, ORDERS};

static void tick_n(int n) {
    for (int i = 0; i < n; i++)
        mus_tick();
}

int main(void) {
    CHECK("minor chord root", _mus_chord_note(129, 0) == 48);
    CHECK("major chord root", _mus_chord_note(136, 0) == 55);

    mus_load(&SONG);
    CHECK("load resets position", mus_pos() == 0);
    mus_play();
    g_ncalls = 0;

    tick_n(2);
    CHECK("row 0 triggers both notes", g_ncalls == 2);
    CHECK("MIDI note passes to host note mode",
          g_calls[0].freq == 69 && (g_calls[0].flags & VEX_TONE_NOTE_MODE));
    CHECK("default volume peaks at level",
          g_calls[0].vol == VEX_TONE_VOLUME(80, 80));
    CHECK("zero sustain defaults to speed*2",
          (g_calls[0].dur & 0xFF) == 2 * 2);
    CHECK("note lands on channel 0", (g_calls[0].flags & 3) == 0);
    CHECK("second note lands on channel 1", (g_calls[1].flags & 3) == 1);
    CHECK("position advances one row", mus_pos() == (1 << 8));

    tick_n(2);
    CHECK("OFF releases the current voice",
          g_ncalls == 3 &&
              g_calls[2].dur == VEX_TONE_DURATION(0, 0, 0, 4) &&
              (g_calls[2].flags & VEX_TONE_RELEASE));

    tick_n(2);
    CHECK("chord and noise trigger", g_ncalls == 5);
    CHECK("chord starts at its root",
          g_calls[3].freq == 48 && (g_calls[3].flags & VEX_TONE_NOTE_MODE));
    CHECK("override volume peaks at override",
          g_calls[3].vol == VEX_TONE_VOLUME(33, 33));
    CHECK("explicit sustain passes through", (g_calls[3].dur & 0xFF) == 7);
    CHECK("noise note uses noise mode", g_calls[4].flags & VEX_TONE_NOISE);

    tick_n(1);
    CHECK("bad instrument skipped and noise OFF hard-cuts",
          g_ncalls == 6 && g_calls[5].dur == 0 && g_calls[5].vol == 0 &&
              !(g_calls[5].flags & VEX_TONE_RELEASE));
    tick_n(1);
    CHECK("end of song stops with 4 silences", g_ncalls == 10);
    tick_n(4);
    CHECK("stopped song stays silent", g_ncalls == 10);

    mus_load(&SONG);
    mus_play();
    g_ncalls = 0;
    CHECK("reload resets position", mus_pos() == 0);
    tick_n(2);
    CHECK("reloaded song triggers again", g_ncalls == 2);

    mus_load(&ARP_SONG);
    mus_play();
    g_ncalls = 0;
    tick_n(14);
    CHECK("chord arpeggio triggers three notes then rests", g_ncalls == 7);
    CHECK("arpeggio starts at root", g_calls[0].freq == 48);
    CHECK("arpeggio reaches third", g_calls[1].freq == 51);
    CHECK("arpeggio reaches fifth", g_calls[2].freq == 55);
    CHECK("arpeggio preserves explicit sustain",
          (g_calls[0].dur & 0xFF) == 7 && (g_calls[1].dur & 0xFF) == 7 &&
              (g_calls[2].dur & 0xFF) == 7);

    mus_load(&HOLD_SONG);
    mus_play();
    g_ncalls = 0;
    tick_n(1);
    CHECK("hold sustain sets hold mode",
          g_ncalls == 2 && (g_calls[0].flags & VEX_TONE_HOLD) &&
              (g_calls[0].dur & 0xFF) == MUS_SUSTAIN_HOLD);
    tick_n(1);
    CHECK("OFF releases held voice", g_ncalls == 8 &&
          (g_calls[2].flags & VEX_TONE_RELEASE));
    CHECK("zero-volume instrument hard-cuts",
          g_calls[3].dur == 0 && g_calls[3].vol == 0 &&
              (g_calls[3].flags & 3) == 1);

    if (failures) {
        fprintf(stderr, "MUS: %d failure(s)\n", failures);
        return 1;
    }
    printf("MUS: all ok\n");
    return 0;
}

File: cmd/vex/test/mus_test.zig
const std = @import("std");
const mus = @import("mus");
const vex = @import("vex");

const instruments = [_]mus.Inst{
    .{
        .wave = vex.TONE_PULSE,
        .duty = vex.TONE_MODE0,
        .attack = 0,
        .decay = 0,
        .sustain = 0,
        .release = 1,
        .volume = 0,
        .pan = 0,
    },
    .{
        .wave = vex.TONE_PULSE,
        .duty = vex.TONE_MODE0,
        .attack = 0,
        .decay = 0,
        .sustain = 0,
        .release = 4,
        .volume = 60,
        .pan = 0,
        .pwm = 1,
        .pwm_start = 32,
        .pwm_end = 192,
    },
};

const events = [_]mus.Event{
    .{ .note = 60, .inst = 2, .vol = 0 },
    .{ .note = 129, .inst = 2, .vol = 0 },
    .{ .note = mus.REST, .inst = 0, .vol = 0 },
    .{ .note = mus.REST, .inst = 0, .vol = 0 },
    .{ .note = 60, .inst = 1, .vol = 0 },
    .{ .note = mus.REST, .inst = 0, .vol = 0 },
    .{ .note = mus.REST, .inst = 0, .vol = 0 },
    .{ .note = mus.REST, .inst = 0, .vol = 0 },
};

const pattern = mus.Pat{ .rows = 2, .speed = 7, .events = &events };
const patterns = [_]*const mus.Pat{&pattern};
const orders = [_]u8{0};
const song = mus.Song{
    .num_insts = instruments.len,
    .num_pats = patterns.len,
    .num_orders = orders.len,
    .loop_ord = 0xFF,
    .insts = &instruments,
    .pats = &patterns,
    .orders = &orders,
};

const fm_instruments = [_]mus.Inst{
    .{
        .wave = vex.TONE_PULSE,
        .duty = vex.TONE_MODE0,
        .attack = 0,
        .decay = 0,
        .sustain = 2,
        .release = 2,
        .volume = 50,
        .pan = 0,
        .fm = 1,
        .fm_ratio = 2,
        .fm_index = 128,
    },
};
const fm_events = [_]mus.Event{
    .{ .note = 139, .inst = 1, .vol = 0 },
    .{ .note = 142, .inst = 1, .vol = 0 },
    .{ .note = mus.REST, .inst = 0, .vol = 0 },
    .{ .note = mus.REST, .inst = 0, .vol = 0 },
};
const fm_pattern = mus.Pat{ .rows = 1, .speed = 7, .events = &fm_events };
const fm_patterns = [_]*const mus.Pat{&fm_pattern};
const fm_orders = [_]u8{0};
const fm_song = mus.Song{
    .num_insts = fm_instruments.len,
    .num_pats = fm_patterns.len,
    .num_orders = fm_orders.len,
    .loop_ord = 0xFF,
    .insts = &fm_instruments,
    .pats = &fm_patterns,
    .orders = &fm_orders,
};
const zero_speed_pattern = mus.Pat{ .rows = 1, .speed = 0, .events = &fm_events };
const zero_speed_patterns = [_]*const mus.Pat{&zero_speed_pattern};
const zero_speed_song = mus.Song{
    .num_insts = fm_instruments.len,
    .num_pats = zero_speed_patterns.len,
    .num_orders = fm_orders.len,
    .loop_ord = 0xFF,
    .insts = &fm_instruments,
    .pats = &zero_speed_patterns,
    .orders = &fm_orders,
};

test "pre-load mute is state-only" {
    mus.mute(0, false);
    vex.reset();
    mus.mute(0, true);
    try std.testing.expectEqual(@as(usize, 0), vex.call_count);

    mus.load(&song);
    mus.play();
    vex.reset();
    mus.tick();
    try std.testing.expectEqual(@as(usize, 1), vex.call_count);
    try std.testing.expectEqual(@as(i32, 1), vex.calls[0].flags & 3);
    mus.mute(0, false);
}

test "zero volume, arpeggio, and rest" {
    mus.load(&song);
    mus.play();
    vex.reset();

    for (0..14) |_| mus.tick();

    try std.testing.expect((mus.pos() & 0xFF) == 0);
    try std.testing.expectEqual(@as(usize, 9), vex.call_count);
    try std.testing.expectEqual(@as(i32, 60), vex.calls[0].freq);
    try std.testing.expectEqual(@as(i32, 0), vex.calls[4].duration);
    try std.testing.expectEqual(@as(i32, 0), vex.calls[4].volume);
    try std.testing.expectEqual(@as(i32, 0), vex.calls[4].flags & 3);

    const expected_notes = [_]i32{ 48, 51, 55 };
    for (expected_notes, 1..4) |note, index| {
        try std.testing.expectEqual(note, vex.calls[index].freq);
        try std.testing.expectEqual(@as(i32, 14), vex.calls[index].duration & 0xFF);
        try std.testing.expect(vex.calls[index].flags & vex.TONE_NOTE_MODE != 0);
        try std.testing.expect(vex.calls[index].flags & vex.TONE_PWM != 0);
        try std.testing.expectEqual(@as(i32, 32), (vex.calls[index].flags >> 12) & 255);
        try std.testing.expectEqual(@as(i32, 192), (vex.calls[index].flags >> 20) & 255);
    }

    for (vex.calls[5..9], 0..4) |call, channel| {
        try std.testing.expectEqual(@as(i32, 0), call.duration);
        try std.testing.expectEqual(@as(i32, 0), call.volume);
        try std.testing.expectEqual(@as(i32, @intCast(channel)), call.flags & 3);
    }
}

test "FM instrument" {
    mus.load(&fm_song);
    mus.play();
    vex.reset();
    mus.tick();
    try std.testing.expectEqual(@as(usize, 1), vex.call_count);
    try std.testing.expectEqual(@as(i32, 60), vex.calls[0].freq);
    try std.testing.expect(vex.calls[0].flags & vex.TONE_FM != 0);
    try std.testing.expectEqual(@as(i32, 2), (vex.calls[0].volume >> 16) & 255);
    try std.testing.expectEqual(@as(i32, 128), (vex.calls[0].volume >> 24) & 255);
}

test "zero speed stops safely" {
    mus.load(&zero_speed_song);
    mus.play();
    vex.reset();
    mus.tick();
    try std.testing.expectEqual(@as(usize, 4), vex.call_count);
}

test "release ignores frequency" {
    mus.load(&fm_song);
    mus.play();
    vex.reset();
    mus.tick();
    mus.mute(0, true);
    try std.testing.expectEqual(@as(usize, 2), vex.call_count);
    try std.testing.expectEqual(@as(i32, 0), vex.calls[1].freq);
    try std.testing.expect(vex.calls[1].flags & vex.TONE_RELEASE != 0);
    mus.mute(0, false);
}

test "repeated mute preserves active release" {
    mus.mute(0, false);
    mus.load(&song);
    mus.play();
    vex.reset();

    mus.tick();
    try std.testing.expectEqual(@as(usize, 2), vex.call_count);
    mus.mute(0, true);
    mus.mute(0, true);
    try std.testing.expectEqual(@as(usize, 3), vex.call_count);
    try std.testing.expectEqual(@as(i32, 0), vex.calls[2].freq);
    try std.testing.expect(vex.calls[2].flags & vex.TONE_RELEASE != 0);
    mus.mute(0, false);
}

test "mute" {
    mus.mute(0, false);
    mus.load(&song);
    mus.play();
    vex.reset();

    mus.mute(0, true);
    try std.testing.expectEqual(@as(usize, 1), vex.call_count);
    mus.tick();
    try std.testing.expectEqual(@as(usize, 2), vex.call_count);
    try std.testing.expectEqual(@as(i32, 1), vex.calls[1].flags & 3);

    mus.mute(0, false);
    mus.load(&song);
    mus.play();
    vex.reset();
    mus.tick();
    try std.testing.expectEqual(@as(usize, 2), vex.call_count);
    try std.testing.expectEqual(@as(i32, 0), vex.calls[0].flags & 3);
    try std.testing.expectEqual(@as(i32, 1), vex.calls[1].flags & 3);
}
